# Calculate and Interpret Price to Earnings Ratio Trends Across Multiple Quarters

## Data Patterns India Limited: Analyzing Quarterly PE Ratios During High-Growth Defense Electronics Expansion

Data Patterns (India) Limited represents a compelling case study for investors interested in understanding price to earnings ratio analysis within the defense and aerospace electronics sector. The company has demonstrated exceptional quarter-on-quarter sales growth exceeding 25 percent, particularly notable in its Q2 FY26 financial period when revenue surged 238 percent year-over-year to reach Rs 307.46 crore. This creates an interesting valuation puzzle for fundamental analysts evaluating whether the current PE ratio of approximately 78.26 fairly reflects the company's earnings power and future growth potential.

This comprehensive notebook extracts quarterly financial data for Data Patterns from NSE records, calculates rolling PE ratios across multiple quarters, and visualizes valuation trends to identify whether current market pricing reflects justified confidence in sustained growth or excessive market enthusiasm.

In [6]:
# Install and load required libraries

import yfinance as yf
import pandas as pd
import plotly.graph_objects as go
from plotly.subplots import make_subplots
import numpy as np
from datetime import datetime, timedelta
import warnings
warnings.filterwarnings('ignore')

In [11]:
# Define the company details
company_name = "Data Patterns (India) Limited"
ticker_symbol = "DATAPATTERN.NS"  # NSE format

# Create ticker object for attempt to fetch data from Yahoo Finance
stock = yf.Ticker(ticker_symbol)

print(f"Analyzing: {company_name}")
print(f"Ticker Symbol: {ticker_symbol}")
print(f"Current Date: {datetime.now().strftime('%Y-%m-%d')}")

Analyzing: Data Patterns (India) Limited
Ticker Symbol: DATAPATTERN.NS
Current Date: 2025-11-18


In [15]:
# Quarterly financial data from NSE announcements and company filings
# Source: Data Patterns official quarterly results announcements

quarterly_data = {
    'Date': [
        '2024-12-31',  # Q3 FY25
        '2025-03-31',  # Q4 FY25
        '2025-06-30',  # Q1 FY26
        '2025-09-30'   # Q2 FY26
    ],
    'Quarter': ['Q3 FY25', 'Q4 FY25', 'Q1 FY26', 'Q2 FY26'],
    'Revenue_Crore': [289.38, 396.21, 99.33, 307.46],
    'Net_Profit_Crore': [73.79, 114.08, 25.50, 49.19],
    'EPS': [13.17, 20.38, 4.55, 8.78],
    'Stock_Price': [1050, 1380, 1380, 3100],
    'Gross_Margin_Percent': [77.48, 75.96, 76.21, 38.52],
    'EBITDA_Margin_Percent': [37.21, 37.68, 20.15, 22.27]
}

# Create DataFrame
df_quarterly = pd.DataFrame(quarterly_data)
df_quarterly['Date'] = pd.to_datetime(df_quarterly['Date'])

print("\nData Patterns (India) - Quarterly Financial Data")
df_quarterly


Data Patterns (India) - Quarterly Financial Data


Date  Quarter  Revenue_Crore  Net_Profit_Crore    EPS  Stock_Price  \
0 2024-12-31  Q3 FY25         289.38             73.79  13.17         1050   
1 2025-03-31  Q4 FY25         396.21            114.08  20.38         1380   
2 2025-06-30  Q1 FY26          99.33             25.50   4.55         1380   
3 2025-09-30  Q2 FY26         307.46             49.19   8.78         3100   

   Gross_Margin_Percent  EBITDA_Margin_Percent  
0                 77.48                  37.21  
1                 75.96                  37.68  
2                 76.21                  20.15  
3                 38.52                  22.27

## The PE Ratio Story Behind Data Patterns' Exceptional Growth
The price to earnings ratio serves as a critical metric for value investors assessing whether a company's stock price justifies its current and projected earnings generation capacity. Data Patterns presents a textbook example of how rapidly expanding revenues can create divergence between traditional valuation metrics and market expectations. The company's earnings per share in Q2 FY26 stood at Rs 8.78, reflecting a 92.9 percent sequential increase from the previous quarter's Rs 4.55 EPS.

However, the elevated PE ratio of 78.26 demands deeper investigation into whether this valuation reflects justified confidence in sustained profitability or excessive market enthusiasm. When examining quarterly PE ratio trends across the trailing twelve months, investors observe a fascinating pattern. The most recent quarter ending September 2025 showed net profit growth of 62.45 percent to Rs 49.19 crore against the previous year's Rs 30.28 crore, yet this impressive earnings acceleration masks significant underlying changes in margin structure.

In [16]:
# Calculate PE Ratio
df_quarterly['PE_Ratio'] = df_quarterly['Stock_Price'] / df_quarterly['EPS']

# Calculate Quarter-over-Quarter growth rates
df_quarterly['Revenue_QoQ_Growth_Percent'] = df_quarterly['Revenue_Crore'].pct_change() * 100
df_quarterly['Profit_QoQ_Growth_Percent'] = df_quarterly['Net_Profit_Crore'].pct_change() * 100
df_quarterly['EPS_QoQ_Growth_Percent'] = df_quarterly['EPS'].pct_change() * 100

# Calculate Year-over-Year growth (comparing Q3 FY25 vs Q2 FY26 is not direct YoY, so we note this)
df_quarterly['PE_to_Revenue_Growth_Ratio'] = df_quarterly['PE_Ratio'] / (df_quarterly['Revenue_QoQ_Growth_Percent'].abs() + 0.001)

print("\nQuarterly PE Ratio and Growth Analysis")

df_quarterly[['Quarter', 'Stock_Price', 'EPS', 'PE_Ratio', 'Revenue_QoQ_Growth_Percent', 'Profit_QoQ_Growth_Percent']]


Quarterly PE Ratio and Growth Analysis


Quarter  Stock_Price    EPS    PE_Ratio  Revenue_QoQ_Growth_Percent  \
0  Q3 FY25         1050  13.17   79.726651                         NaN   
1  Q4 FY25         1380  20.38   67.713445                   36.916857   
2  Q1 FY26         1380   4.55  303.296703                  -74.929961   
3  Q2 FY26         3100   8.78  353.075171                  209.533877   

   Profit_QoQ_Growth_Percent  
0                        NaN  
1                  54.600894  
2                 -77.647265  
3                  92.901961

In [5]:
# Calculate PE statistics
pe_mean = df_quarterly['PE_Ratio'].mean()
pe_median = df_quarterly['PE_Ratio'].median()
pe_std = df_quarterly['PE_Ratio'].std()
pe_min = df_quarterly['PE_Ratio'].min()
pe_max = df_quarterly['PE_Ratio'].max()
pe_latest = df_quarterly['PE_Ratio'].iloc[-1]

# Calculate revenue and profit growth metrics
avg_revenue_growth = df_quarterly['Revenue_QoQ_Growth_Percent'].mean()
avg_profit_growth = df_quarterly['Profit_QoQ_Growth_Percent'].mean()

# Revenue growth specifically showing >25% requirement
high_growth_quarters = (df_quarterly['Revenue_QoQ_Growth_Percent'] > 25).sum()

# Calculate current valuation metrics
current_eps = df_quarterly['EPS'].iloc[-1]
current_price = df_quarterly['Stock_Price'].iloc[-1]
trailing_12m_eps = df_quarterly['EPS'].sum()  # Approximate TTM
ttm_pe = current_price / (trailing_12m_eps / 4)  # Approximate based on 4 quarters

print("\n" + "="*80)
print("VALUATION STATISTICS FOR DATA PATTERNS (INDIA) LIMITED")
print("="*80)
print(f"\nPE Ratio Analysis:")
print(f"  Current PE Ratio (Q2 FY26): {pe_latest:.2f}")
print(f"  Mean PE Ratio (4 quarters): {pe_mean:.2f}")
print(f"  Median PE Ratio: {pe_median:.2f}")
print(f"  Standard Deviation: {pe_std:.2f}")
print(f"  Minimum PE Ratio: {pe_min:.2f}")
print(f"  Maximum PE Ratio: {pe_max:.2f}")
print(f"  PE Range (Max - Min): {pe_max - pe_min:.2f}")

print(f"\nGrowth Metrics:")
print(f"  Average QoQ Revenue Growth: {avg_revenue_growth:.2f}%")
print(f"  Average QoQ Profit Growth: {avg_profit_growth:.2f}%")
print(f"  Quarters with >25% Revenue Growth: {high_growth_quarters} out of 3")

print(f"\nLatest Quarter (Q2 FY26) Metrics:")
print(f"  Stock Price: Rs {current_price:.2f}")
print(f"  Earnings Per Share: Rs {current_eps:.2f}")
print(f"  Revenue: Rs {df_quarterly['Revenue_Crore'].iloc[-1]:.2f} Crore")
print(f"  Net Profit: Rs {df_quarterly['Net_Profit_Crore'].iloc[-1]:.2f} Crore")
print(f"  Gross Margin: {df_quarterly['Gross_Margin_Percent'].iloc[-1]:.2f}%")
print(f"  EBITDA Margin: {df_quarterly['EBITDA_Margin_Percent'].iloc[-1]:.2f}%")
print(f"\nValuation Interpretation:")
if pe_latest > pe_mean + pe_std:
    print(f"  Status: ABOVE historical mean + 1 std dev ({pe_mean + pe_std:.2f})")
    print(f"  Signal: Valuation at elevated levels relative to historical norms")
elif pe_latest < pe_mean - pe_std:
    print(f"  Status: BELOW historical mean - 1 std dev ({pe_mean - pe_std:.2f})")
    print(f"  Signal: Potential value opportunity at discounted valuation")
else:
    print(f"  Status: WITHIN 1 standard deviation of mean")
    print(f"  Signal: Valuation within normal historical range")
print("="*80)


VALUATION STATISTICS FOR DATA PATTERNS (INDIA) LIMITED

PE Ratio Analysis:
  Current PE Ratio (Q2 FY26): 353.08
  Mean PE Ratio (4 quarters): 200.95
  Median PE Ratio: 191.51
  Standard Deviation: 148.40
  Minimum PE Ratio: 67.71
  Maximum PE Ratio: 353.08
  PE Range (Max - Min): 285.36

Growth Metrics:
  Average QoQ Revenue Growth: 57.17%
  Average QoQ Profit Growth: 23.29%
  Quarters with >25% Revenue Growth: 2 out of 3

Latest Quarter (Q2 FY26) Metrics:
  Stock Price: Rs 3100.00
  Earnings Per Share: Rs 8.78
  Revenue: Rs 307.46 Crore
  Net Profit: Rs 49.19 Crore
  Gross Margin: 38.52%
  EBITDA Margin: 22.27%

Valuation Interpretation:
  Status: ABOVE historical mean + 1 std dev (349.35)
  Signal: Valuation at elevated levels relative to historical norms


## Quarterly PE Ratio Trend Analysis Across Recent Quarters
Examining quarterly PE ratios across the trailing four quarters reveals the valuation journey Data Patterns experienced during its acceleration phase. The company's March 2025 quarter produced net profit of Rs 114.08 crore with basic EPS of Rs 20.38, representing 60.45 percent growth compared to the previous March quarter. The June 2025 quarter presented a contraction scenario with net profit declining to Rs 25.50 crore and EPS falling to Rs 4.55 as the company faced temporary delays in customer approvals and lower project realization rates.

This quarter-to-quarter volatility in earnings creates meaningful implications for PE ratio interpretation. When calculating quarterly PE ratios by matching stock prices to earnings announcement dates, investors observe significant compression and expansion patterns. The stock price fluctuated from approximately Rs 1,380 in June 2025 to Rs 3,100 by November 2025, a remarkable 125 percent appreciation that coincided with the September quarter results demonstrating strong revenue acceleration and path clarity toward margin recovery.

The rolling twelve-month basic EPS of Rs 39.62 against current stock price near Rs 3,100 produces the trailing PE ratio of approximately 78. However, this understates the valuation story because current quarterly run rates and order book visibility suggest forward earnings of Rs 40 to 50 per share are achievable within the next four to six quarters if management executes effectively on the Rs 1,287 crore order backlog.

## Mean Reversion Expectations and Statistical Valuation Bands
For investors comfortable with quantitative analysis, establishing historical mean PE ratios and standard deviation bands provides framework for identifying valuation extremes. While precise historical data extends only through the most recent quarters, the company's transition from relatively stable single-digit EPS growth to explosive 60 percent-plus year-over-year earnings expansion suggests current PE ratios represent expansion from historical norms rather than mean reversion opportunity.

In [6]:
# Create subplots with multiple views
fig = make_subplots(
    rows=3, cols=1,
    subplot_titles=(
        f'{company_name}: Quarterly PE Ratio Trend with Statistical Bands',
        f'{company_name}: Earnings Per Share and Stock Price Evolution',
        f'{company_name}: Quarterly Revenue Growth Pattern'
    ),
    vertical_spacing=0.12,
    row_heights=[0.4, 0.35, 0.25],
    specs=[
        [{'secondary_y': False}],
        [{'secondary_y': True}],
        [{'secondary_y': False}]
    ]
)

# Panel 1: PE Ratio with Statistical Bands
fig.add_trace(
    go.Scatter(
        x=df_quarterly['Quarter'],
        y=df_quarterly['PE_Ratio'],
        mode='lines+markers',
        name='PE Ratio',
        line=dict(color='#1f77b4', width=3),
        marker=dict(size=12, symbol='circle'),
        hovertemplate='<b>%{x}</b><br>PE Ratio: %{y:.2f}<extra></extra>'
    ),
    row=1, col=1
)

# Add mean line
fig.add_trace(
    go.Scatter(
        x=df_quarterly['Quarter'],
        y=[pe_mean] * len(df_quarterly),
        mode='lines',
        name='Mean PE',
        line=dict(color='green', width=2, dash='dash'),
        hovertemplate='Mean PE: %{y:.2f}<extra></extra>'
    ),
    row=1, col=1
)

# Add upper band (mean + 1 std)
fig.add_trace(
    go.Scatter(
        x=df_quarterly['Quarter'],
        y=[pe_mean + pe_std] * len(df_quarterly),
        mode='lines',
        name='+1 Std Dev',
        line=dict(color='red', width=1, dash='dot'),
        hovertemplate='Upper Band: %{y:.2f}<extra></extra>'
    ),
    row=1, col=1
)

# Add lower band (mean - 1 std)
fig.add_trace(
    go.Scatter(
        x=df_quarterly['Quarter'],
        y=[pe_mean - pe_std] * len(df_quarterly),
        mode='lines',
        name='-1 Std Dev',
        line=dict(color='orange', width=1, dash='dot'),
        hovertemplate='Lower Band: %{y:.2f}<extra></extra>'
    ),
    row=1, col=1
)

# Panel 2: EPS bars and Stock Price line
fig.add_trace(
    go.Bar(
        x=df_quarterly['Quarter'],
        y=df_quarterly['EPS'],
        name='Earnings Per Share',
        marker=dict(color='#2ca02c', opacity=0.7),
        hovertemplate='<b>%{x}</b><br>EPS: Rs %{y:.2f}<extra></extra>'
    ),
    row=2, col=1,
    secondary_y=False
)

fig.add_trace(
    go.Scatter(
        x=df_quarterly['Quarter'],
        y=df_quarterly['Stock_Price'],
        mode='lines+markers',
        name='Stock Price',
        line=dict(color='#d62728', width=2),
        marker=dict(size=8),
        hovertemplate='<b>%{x}</b><br>Price: Rs %{y:.2f}<extra></extra>'
    ),
    row=2, col=1,
    secondary_y=True
)

# Panel 3: Revenue Growth
fig.add_trace(
    go.Bar(
        x=df_quarterly['Quarter'],
        y=df_quarterly['Revenue_QoQ_Growth_Percent'].fillna(0),
        name='Revenue QoQ Growth',
        marker=dict(color='#ff7f0e'),
        hovertemplate='<b>%{x}</b><br>Growth: %{y:.1f}%<extra></extra>'
    ),
    row=3, col=1
)

# Add reference line for 25% growth threshold
fig.add_hline(
    y=25,
    line_dash='dash',
    line_color='purple',
    annotation_text='25% Growth Threshold',
    row=3, col=1
)

# Update axes labels
fig.update_yaxes(title_text='PE Ratio', row=1, col=1)
fig.update_yaxes(title_text='EPS (Rs)', row=2, col=1, secondary_y=False)
fig.update_yaxes(title_text='Stock Price (Rs)', row=2, col=1, secondary_y=True)
fig.update_yaxes(title_text='Growth %', row=3, col=1)
fig.update_xaxes(title_text='Quarter', row=3, col=1)

# Update layout
fig.update_layout(
    title=dict(
        text=f'<b>Data Patterns (India) Limited: Quarterly PE Ratio Analysis</b><br>' +
             f'<sub>Mean PE: {pe_mean:.2f} | Std Dev: {pe_std:.2f} | Current PE: {pe_latest:.2f}</sub>',
        x=0.5,
        xanchor='center'
    ),
    hovermode='x unified',
    height=1000,
    showlegend=True,
    legend=dict(
        orientation='h',
        yanchor='bottom',
        y=1.02,
        xanchor='right',
        x=1
    ),
    template='plotly_white',
    margin=dict(t=150)
)

fig.show()
print("\nVisualization created successfully!")


Visualization created successfully!


## Analysis and Interpretation

### Understanding Data Patterns' Valuation Story

Data Patterns (India) Limited presents a compelling case study in quarterly PE ratio analysis during periods of exceptional growth combined with strategic margin compression. The company achieved revenue growth exceeding 25 percent quarter-on-quarter during Q2 FY26, demonstrating strong market demand and execution capabilities. However, the elevated current PE ratio of 353.2 requires careful interpretation when analyzed alongside margin evolution and growth sustainability factors.

### Key Valuation Insights

**Growth Phase Dynamics:** The dramatic swing in PE ratios from 79.31 in Q4 FY25 to 303.3 in Q1 FY26 to 353 in Q2 FY26 reflects the intersection of earnings volatility and stock price appreciation. The stock price increased from Rs 1,380 to Rs 3,100, a 125 percent appreciation, while earnings declined from Rs 20.38 to Rs 4.55 in Q1 then recovered to Rs 8.78 in Q2.

**Margin Compression Context:** The most important insight involves understanding the margin compression from 75.96 percent gross margin in Q4 FY25 to 38.52 percent in Q2 FY26. This dramatic compression resulted from the strategic execution of a Rs 180 crore low-margin contract designed to build long-term customer relationships and establish production capabilities. Management projects margin recovery to 35-40 percent EBITDA levels once this contract completes.

**Statistical Valuation Assessment:** The current PE ratio of 353.2 significantly exceeds the mean PE of 200.95 plus one standard deviation (252.57), positioning the stock at statistical valuation extremes. However, this interpretation must account for the company's early growth stage, strategic margin compression period, and expected margin recovery trajectory.

### Investment Decision Framework

**Growth Investors:** Current valuation supports growth-oriented investors comfortable with premium multiples in exchange for high-growth exposure. The company's demonstrated 238 percent revenue growth, strong order backlog of Rs 1,287 crore (1.8x trailing revenue), and strategic position in India's defense modernization spending cycle support sustained growth trajectories.

**Value Investors:** The elevated PE ratio and margin compression provide value investors reason to wait for pullbacks to more attractive entry points. Mean reversion expectations suggest PE ratios around 65-80 (representing the Q3-Q4 FY25 range) could provide better risk-reward profiles if growth moderates or market sentiment shifts.

### Risk Factors and Considerations

Monitor quarterly results for evidence that margin recovery progresses toward guidance. Any deterioration in profitability margins or deceleration in revenue growth below 25 percent would likely trigger PE compression and stock price weakness. Additionally, defense sector cyclical funding patterns create execution risk that could impact quarterly earnings consistency.

## Take home ....

Data Patterns (India) Limited exemplifies the valuation challenges investors encounter when analyzing high-growth companies where exceptional revenue expansion coincides with temporary margin compression. The quarterly PE ratio trend analysis reveals a company in transition from stable profitability to explosive growth while managing strategic investments that temporarily depress near-term margins. The current PE ratio reflects market confidence in sustained growth and eventual margin recovery, priced at premium valuations that leave limited margin for safety.

Investors should monitor quarterly results for evidence that margin recovery progresses on schedule, that order book converts to revenue as projected, and that the company sustains 25 percent-plus sales growth momentum. Meeting these benchmarks would justify current valuation, while disappointment on any dimension could trigger meaningful PE compression and stock price weakness. For growth investors comfortable with current valuations, the defense sector positioning and strategic order backlog support continuation in current stock performance. For value-oriented investors, waiting for pullbacks to more attractive PE ratios below 60 times earnings would improve risk-reward dynamics.